torch/utils/data/dataloader.py

```diff
@@ -55,131 +81,84 @@ class DataLoader(object):
-            return SimpleCustomBatch(batch)
-
-        inps = torch.arange(10 * 5, dtype=torch.float32).view(10, 5)
-        tgts = torch.arange(10 * 5, dtype=torch.float32).view(10, 5)
-        dataset = TensorDataset(inps, tgts)
-
-        loader = DataLoader(dataset, batch_size=2, collate_fn=collate_wrapper,
-                            pin_memory=True)
-
-        for batch_ndx, sample in enumerate(loader):
-            print(sample.inp.is_pinned())
-            print(sample.tgt.is_pinned())
+    .. note:: When :attr:`dataset` is an :class:`~torch.utils.data.IterableDataset`,
+              ``len(dataloader)`` returns ``len(dataset)``. But this may not be
+              accurate because the actual iterator size also depends on :attr:`num_workers`
+              and :attr:`worker_init_fn`. If ``num_workers > 0``, each worker
+              gets a copy of the same iterable dataset object and can return
+              duplicate data, unless the dataset copies and/or the workers are
+              configured differently in :attr:`worker_init_fn`. See
+              :class:`~torch.utils.data.IterableDataset` for more details and
+              examples.
 
     """
 
     __initialized = False
 
     def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
-                 batch_sampler=None, num_workers=0, collate_fn=default_collate,
+                 batch_sampler=None, num_workers=0,
+                 convert_fn=_utils.collate.default_convert,
+                 collate_fn=default_collate,
                  pin_memory=False, drop_last=False, timeout=0,
                  worker_init_fn=None):
         self.dataset = dataset
-        self.batch_size = batch_size
         self.num_workers = num_workers
-        self.collate_fn = collate_fn
         self.pin_memory = pin_memory
-        self.drop_last = drop_last
         self.timeout = timeout
         self.worker_init_fn = worker_init_fn
-
-        if timeout < 0:
-            raise ValueError('timeout option should be non-negative')
-
-        if batch_sampler is not None:
-            if batch_size > 1 or shuffle or sampler is not None or drop_last:
-                raise ValueError('batch_sampler option is mutually exclusive '
-                                 'with batch_size, shuffle, sampler, and '
-                                 'drop_last')
-            self.batch_size = None
-            self.drop_last = None
-
-        if sampler is not None and shuffle:
-            raise ValueError('sampler option is mutually exclusive with '
-                             'shuffle')
-
-        if self.num_workers < 0:
-            raise ValueError('num_workers option cannot be negative; '
-                             'use num_workers=0 to disable multiprocessing.')
-
-        if batch_sampler is None:
+        self.convert_fn = convert_fn
+        self.collate_fn = collate_fn
+        if isinstance(dataset, IterableDataset):
+            # FIXME: check default for some args
```

Comment:
Please do! At least check that the sampler is `None`. We should do this before we land this PR, otherwise it's pretty bad UX